File: commonplace_log/test/merge_plan_test.exs
defmodule Commonplace.Log.MergePlanTest.Fixtures do
  @moduledoc false
  # Fixture builders shared by the compile-time table and the runtime tests.

  # Deterministic entry maker: the fake canonical bytes encode every argument,
  # so two calls with identical arguments are byte-equal and any argument
  # difference (including payload alone, for collision/fork rows) changes the
  # bytes. IDs are opaque strings to the classifier; no need for real UUIDs.
  def mk(writer_id, writer_seq, entry_id, prev_entry_id, payload \\ "") do
    %{
      entry_id: entry_id,
      writer_id: writer_id,
      writer_seq: writer_seq,
      prev_entry_id: prev_entry_id,
      canonical_bytes:
        "#{writer_id}/#{writer_seq}/#{entry_id}/#{prev_entry_id || "null"}/#{payload}"
    }
  end

  # Store state for a row: entries already present, tips derived from them.
  def store_of(present) do
    existing =
      Enum.reduce(present, %{}, fn e, acc ->
        Map.update(
          acc,
          e.writer_id,
          %{seq: e.writer_seq, entry_id: e.entry_id},
          fn tip ->
            if e.writer_seq > tip.seq,
              do: %{seq: e.writer_seq, entry_id: e.entry_id},
              else: tip
          end
        )
      end)

    present_by_coord = fn w, s ->
      case Enum.find(present, &(&1.writer_id == w and &1.writer_seq == s)) do
        nil -> nil
        e -> e.canonical_bytes
      end
    end

    present_by_id = fn id ->
      case Enum.find(present, &(&1.entry_id == id)) do
        nil ->
          nil

        e ->
          %{writer_id: e.writer_id, writer_seq: e.writer_seq, bytes: e.canonical_bytes}
      end
    end

    {existing, present_by_coord, present_by_id}
  end
end

defmodule Commonplace.Log.MergePlanTest do
  @moduledoc """
  Table-driven tests for the pure merge classifier (spec §9.3 rules 1–6 and
  the three merge error codes, §6.3 entry-ID collision, §8 invariants
  3/4/9/10). Ported row-for-row from `worker/test/merge-plan.test.ts` — the
  two suites must stay in lockstep; a row changed on one side must change on
  the other.
  """
  use ExUnit.Case, async: true

  alias Commonplace.Log.MergePlan
  alias Commonplace.Log.MergePlanTest.Fixtures

  import Fixtures, only: [mk: 4, store_of: 1]

  # --- match helper --------------------------------------------------------

  # Partial structural match, mirroring vitest's `toMatchObject`: maps match
  # by subset (recursively), lists match by length and elementwise, everything
  # else by equality.
  defp matches?(actual, expected) when is_map(expected) and is_map(actual) do
    Enum.all?(expected, fn {k, v} ->
      Map.has_key?(actual, k) and matches?(Map.fetch!(actual, k), v)
    end)
  end

  defp matches?(actual, expected) when is_list(expected) and is_list(actual) do
    length(actual) == length(expected) and
      Enum.zip(actual, expected) |> Enum.all?(fn {a, e} -> matches?(a, e) end)
  end

  defp matches?({:ok, a}, {:ok, e}), do: matches?(a, e)
  defp matches?({:error, {code, a}}, {:error, {code, e}}), do: matches?(a, e)
  defp matches?(actual, expected), do: actual == expected

  defp assert_matches(actual, expected) do
    unless matches?(actual, expected) do
      flunk("""
      expected (partial): #{inspect(expected, pretty: true)}
      actual: #{inspect(actual, pretty: true)}
      """)
    end
  end

  # --- fixtures ------------------------------------------------------------

  # Writer W with three committed entries; the standard "existing store" for
  # most rows. Writer ids chosen so sort order is obvious in multi-writer rows.
  @w "writer-mmm"
  @w1 Fixtures.mk(@w, 1, "id-w1", nil)
  @w2 Fixtures.mk(@w, 2, "id-w2", "id-w1")
  @w3 Fixtures.mk(@w, 3, "id-w3", "id-w2")

  # --- the table -----------------------------------------------------------

  # Each row: {name, present, batch, expected (partial outcome)}. Lockstep
  # with the TS twin's 19-row table, same order, same names (adapted).
  @rows [
    {
      "empty batch is ok with zero plans (§9.3 vacuous case)",
      [@w1, @w2],
      [],
      {:ok, %{per_writer: [], present_count: 0}}
    },
    {
      "fresh writer starting at seq 1 with chained prevs (§9.3 rules 3–5)",
      [],
      [Fixtures.mk("wf", 1, "f1", nil), Fixtures.mk("wf", 2, "f2", "f1")],
      {:ok,
       %{
         present_count: 0,
         per_writer: [
           %{
             writer_id: "wf",
             first_new_seq: 1,
             new_entries: [%{entry_id: "f1"}, %{entry_id: "f2"}]
           }
         ]
       }}
    },
    {
      "fresh writer NOT starting at 1 → writer_gap with expected_seq 1 and nil tip",
      [],
      [Fixtures.mk("wf", 2, "f2", "f1")],
      {:error, {:writer_gap, %{writer_id: "wf", expected_seq: 1, tip_entry_id: nil}}}
    },
    {
      "extension at tip+1 naming the tip in prev_entry_id (§9.3 rules 3–4)",
      [@w1, @w2],
      [Fixtures.mk(@w, 3, "id-w3", "id-w2")],
      {:ok,
       %{
         present_count: 0,
         per_writer: [
           %{writer_id: @w, first_new_seq: 3, new_entries: [%{entry_id: "id-w3"}]}
         ]
       }}
    },
    {
      "extension at tip+1 with wrong prev_entry_id → writer_fork (§9.3 rule 4: coordinate holds a different chain)",
      [@w1, @w2],
      [Fixtures.mk(@w, 3, "id-x3", "id-w1")],
      {:error, {:writer_fork, %{writer_id: @w, seq: 3}}}
    },
    {
      "batch wholly at/below tip and byte-equal → all present, no plan (§9.3 rule 2, invariant 9)",
      [@w1, @w2, @w3],
      [@w1, @w2, @w3],
      {:ok, %{per_writer: [], present_count: 3}}
    },
    {
      "partially duplicate batch: present prefix counted, new suffix planned",
      [@w1, @w2],
      [@w2, Fixtures.mk(@w, 3, "id-w3", "id-w2")],
      {:ok,
       %{
         present_count: 1,
         per_writer: [
           %{writer_id: @w, first_new_seq: 3, new_entries: [%{entry_id: "id-w3"}]}
         ]
       }}
    },
    {
      "first new entry beyond tip+1 → writer_gap carrying the receiver's tip (§9.3 error 1, §15.5)",
      [@w1, @w2],
      [Fixtures.mk(@w, 4, "id-w4", "id-w3")],
      {:error, {:writer_gap, %{writer_id: @w, expected_seq: 3, tip_entry_id: "id-w2"}}}
    },
    {
      "different bytes at an occupied coordinate → writer_fork (§9.3 error 2)",
      [@w1, @w2],
      [Fixtures.mk(@w, 2, "id-x2", "id-w1", "divergent")],
      {:error, {:writer_fork, %{writer_id: @w, seq: 2}}}
    },
    {
      "existing entry_id with different canonical bytes → entry_id_collision (§9.3 error 3, §6.3)",
      [@w1, @w2],
      # A structurally valid extension that reuses id-w1 for different bytes.
      [Fixtures.mk(@w, 3, "id-w1", "id-w2", "reused-id")],
      {:error, {:entry_id_collision, %{entry_id: "id-w1"}}}
    },
    {
      "same entry_id at the SAME coordinate with different bytes → entry_id_collision, not fork (§6.3 MUST)",
      [@w1, @w2],
      # Same id, same (writer, seq) as the stored w2, but tampered bytes.
      # Kills the mutant that drops the byte-equality clause from the
      # stored-id check: without it this row degrades to writer_fork.
      [Fixtures.mk(@w, 2, "id-w2", "id-w1", "tampered")],
      {:error, {:entry_id_collision, %{entry_id: "id-w2"}}}
    },
    {
      "same entry_id, same bytes, same coordinate → duplicate, present, NOT a collision (§9.3 rule 2)",
      [@w1, @w2],
      [Fixtures.mk(@w, 2, "id-w2", "id-w1")],
      {:ok, %{per_writer: [], present_count: 1}}
    },
    {
      "intra-batch sequence hole after a valid first new entry → writer_gap at the hole (§9.3 rule 5)",
      [@w1, @w2],
      [
        Fixtures.mk(@w, 3, "id-w3", "id-w2"),
        Fixtures.mk(@w, 4, "id-w4", "id-w3"),
        Fixtures.mk(@w, 6, "id-w6", "id-w5")
      ],
      {:error, {:writer_gap, %{writer_id: @w, expected_seq: 5, tip_entry_id: "id-w2"}}}
    },
    {
      "intra-batch prev_entry_id not naming the preceding batch entry → invalid_batch (§9.3 rule 5)",
      [@w1, @w2],
      [Fixtures.mk(@w, 3, "id-w3", "id-w2"), Fixtures.mk(@w, 4, "id-w4", "id-BOGUS")],
      {:error, {:invalid_batch, %{}}}
    },
    {
      "intra-batch entry_id reused at different coordinates → entry_id_collision (§6.3)",
      [],
      [Fixtures.mk("wf", 1, "f-dup", nil), Fixtures.mk("wf", 2, "f-dup", "f-dup")],
      {:error, {:entry_id_collision, %{entry_id: "f-dup"}}}
    },
    {
      "intra-batch duplicate coordinate with different bytes → writer_fork",
      [],
      [Fixtures.mk("wf", 1, "f1", nil), Fixtures.mk("wf", 1, "f1b", nil)],
      {:error, {:writer_fork, %{writer_id: "wf", seq: 1}}}
    },
    {
      "exact duplicate entry within one batch → deduplicated, copy counted as present",
      [],
      [Fixtures.mk("wf", 1, "f1", nil), Fixtures.mk("wf", 1, "f1", nil)],
      {:ok,
       %{
         present_count: 1,
         per_writer: [
           %{writer_id: "wf", first_new_seq: 1, new_entries: [%{entry_id: "f1"}]}
         ]
       }}
    },
    {
      "multi-writer batch where one writer fails → whole batch fails with that writer's error, no partial plan",
      [@w1, @w2],
      [
        # "aaa" sorts before W and is entirely valid; the classifier must
        # still refuse the whole batch for W's gap. An error outcome carries
        # no per_writer, so no partial plan can escape (§8 invariant 10).
        Fixtures.mk("aaa", 1, "a1", nil),
        Fixtures.mk(@w, 4, "id-w4", "id-w3")
      ],
      {:error, {:writer_gap, %{writer_id: @w, expected_seq: 3, tip_entry_id: "id-w2"}}}
    },
    {
      "multi-writer all-valid batch → one plan per writer, sorted by writer_id",
      [@w1, @w2],
      [
        # Deliberately fed in reverse writer order.
        Fixtures.mk("zzz", 1, "z1", nil),
        Fixtures.mk(@w, 3, "id-w3", "id-w2"),
        Fixtures.mk("aaa", 1, "a1", nil),
        Fixtures.mk("aaa", 2, "a2", "a1")
      ],
      {:ok,
       %{
         present_count: 0,
         per_writer: [
           %{
             writer_id: "aaa",
             first_new_seq: 1,
             new_entries: [%{entry_id: "a1"}, %{entry_id: "a2"}]
           },
           %{writer_id: @w, first_new_seq: 3, new_entries: [%{entry_id: "id-w3"}]},
           %{writer_id: "zzz", first_new_seq: 1, new_entries: [%{entry_id: "z1"}]}
         ]
       }}
    }
  ]

  # --- runner --------------------------------------------------------------

  describe "plan_merge classifies batches per spec §9.3" do
    test "anti-vacuity: the table has at least 14 rows" do
      assert length(@rows) >= 14
    end

    # The TS twin has 19 table rows; the tables must not drift apart.
    test "lockstep: the table has exactly the TS twin's 19 rows" do
      assert length(@rows) == 19
    end

    for {{name, present, batch, expected}, index} <- Enum.with_index(@rows) do
      @row {present, batch, expected}
      test "row #{index}: #{name}" do
        {present, batch, expected} = @row
        {existing, present_by_coord, present_by_id} = store_of(present)
        outcome = MergePlan.plan_merge(batch, existing, present_by_coord, present_by_id)
        assert_matches(outcome, expected)
      end
    end

    test "batch order does not matter: shuffled input classifies identically (§9.3 rule 1)" do
      batch = [
        mk(@w, 3, "id-w3", "id-w2"),
        mk(@w, 4, "id-w4", "id-w3"),
        mk(@w, 5, "id-w5", "id-w4")
      ]

      shuffled = [Enum.at(batch, 2), Enum.at(batch, 0), Enum.at(batch, 1)]
      {existing, present_by_coord, present_by_id} = store_of([@w1, @w2])
      a = MergePlan.plan_merge(batch, existing, present_by_coord, present_by_id)
      b = MergePlan.plan_merge(shuffled, existing, present_by_coord, present_by_id)
      assert b == a

      # Pinned non-trivial shape so both runs can't agree on garbage.
      assert_matches(
        a,
        {:ok,
         %{
           present_count: 0,
           per_writer: [
             %{
               writer_id: @w,
               first_new_seq: 3,
               new_entries: [
                 %{entry_id: "id-w3"},
                 %{entry_id: "id-w4"},
                 %{entry_id: "id-w5"}
               ]
             }
           ]
         }}
      )
    end
  end
end

File: worker/test/merge-plan.test.ts
/**
 * Table-driven tests for the pure merge classifier (spec §9.3 rules 1–6 and
 * the three merge error codes, §6.3 entry-ID collision, §8 invariants
 * 3/4/9/10). The module under test is pure — no database, no I/O — so this
 * suite runs in the plain-node unit project.
 */
import { describe, expect, it } from "vitest";
import {
  planMerge,
  type BatchEntry,
  type MergeOutcome,
  type TipInfo,
} from "../src/do/merge-plan";

// --- fixtures --------------------------------------------------------------

const encoder = new TextEncoder();

/**
 * Deterministic entry maker: the fake canonical bytes encode every argument,
 * so two calls with identical arguments are byte-equal and any argument
 * difference (including payload alone, for collision/fork rows) changes the
 * bytes. IDs are opaque strings to the classifier; no need for real UUIDs.
 */
function mk(
  writerId: string,
  writerSeq: number,
  entryId: string,
  prevEntryId: string | null,
  payload = "",
): BatchEntry {
  return {
    entryId,
    writerId,
    writerSeq,
    prevEntryId,
    canonicalBytes: encoder.encode(
      `${writerId}/${writerSeq}/${entryId}/${prevEntryId}/${payload}`,
    ),
  };
}

/** Store state for a row: entries already present, tips derived from them. */
function storeOf(present: BatchEntry[]): {
  existing: Map<string, TipInfo>;
  presentByCoord: (w: string, s: number) => Uint8Array | null;
  presentById: (
    id: string,
  ) => { writerId: string; writerSeq: number; bytes: Uint8Array } | null;
} {
  const existing = new Map<string, TipInfo>();
  for (const e of present) {
    const tip = existing.get(e.writerId);
    if (!tip || e.writerSeq > tip.seq) {
      existing.set(e.writerId, { seq: e.writerSeq, entryId: e.entryId });
    }
  }
  return {
    existing,
    presentByCoord: (w, s) =>
      present.find((e) => e.writerId === w && e.writerSeq === s)
        ?.canonicalBytes ?? null,
    presentById: (id) => {
      const e = present.find((p) => p.entryId === id);
      return e
        ? {
            writerId: e.writerId,
            writerSeq: e.writerSeq,
            bytes: e.canonicalBytes,
          }
        : null;
    },
  };
}

// Writer W with three committed entries; the standard "existing store" for
// most rows. Writer ids chosen so sort order is obvious in multi-writer rows.
const W = "writer-mmm";
const w1 = mk(W, 1, "id-w1", null);
const w2 = mk(W, 2, "id-w2", "id-w1");
const w3 = mk(W, 3, "id-w3", "id-w2");

// --- the table -------------------------------------------------------------

interface Row {
  name: string;
  present: BatchEntry[];
  batch: BatchEntry[];
  /** Partial structural match against the outcome (arrays match by length). */
  expected: Record<string, unknown>;
}

const rows: Row[] = [
  {
    name: "empty batch is ok with zero plans (§9.3 vacuous case)",
    present: [w1, w2],
    batch: [],
    expected: { ok: true, perWriter: [], presentCount: 0 },
  },
  {
    name: "fresh writer starting at seq 1 with chained prevs (§9.3 rules 3–5)",
    present: [],
    batch: [mk("wf", 1, "f1", null), mk("wf", 2, "f2", "f1")],
    expected: {
      ok: true,
      presentCount: 0,
      perWriter: [
        {
          writerId: "wf",
          firstNewSeq: 1,
          newEntries: [{ entryId: "f1" }, { entryId: "f2" }],
        },
      ],
    },
  },
  {
    name: "fresh writer NOT starting at 1 → writer_gap with expectedSeq 1 and null tip",
    present: [],
    batch: [mk("wf", 2, "f2", "f1")],
    expected: {
      ok: false,
      code: "writer_gap",
      details: { writerId: "wf", expectedSeq: 1, tipEntryId: null },
    },
  },
  {
    name: "extension at tip+1 naming the tip in prev_entry_id (§9.3 rules 3–4)",
    present: [w1, w2],
    batch: [mk(W, 3, "id-w3", "id-w2")],
    expected: {
      ok: true,
      presentCount: 0,
      perWriter: [
        { writerId: W, firstNewSeq: 3, newEntries: [{ entryId: "id-w3" }] },
      ],
    },
  },
  {
    name: "extension at tip+1 with wrong prev_entry_id → writer_fork (§9.3 rule 4: coordinate holds a different chain)",
    present: [w1, w2],
    batch: [mk(W, 3, "id-x3", "id-w1")],
    expected: {
      ok: false,
      code: "writer_fork",
      details: { writerId: W, seq: 3 },
    },
  },
  {
    name: "batch wholly at/below tip and byte-equal → all present, no plan (§9.3 rule 2, invariant 9)",
    present: [w1, w2, w3],
    batch: [w1, w2, w3],
    expected: { ok: true, perWriter: [], presentCount: 3 },
  },
  {
    name: "partially duplicate batch: present prefix counted, new suffix planned",
    present: [w1, w2],
    batch: [w2, mk(W, 3, "id-w3", "id-w2")],
    expected: {
      ok: true,
      presentCount: 1,
      perWriter: [
        { writerId: W, firstNewSeq: 3, newEntries: [{ entryId: "id-w3" }] },
      ],
    },
  },
  {
    name: "first new entry beyond tip+1 → writer_gap carrying the receiver's tip (§9.3 error 1, §15.5)",
    present: [w1, w2],
    batch: [mk(W, 4, "id-w4", "id-w3")],
    expected: {
      ok: false,
      code: "writer_gap",
      details: { writerId: W, expectedSeq: 3, tipEntryId: "id-w2" },
    },
  },
  {
    name: "different bytes at an occupied coordinate → writer_fork (§9.3 error 2)",
    present: [w1, w2],
    batch: [mk(W, 2, "id-x2", "id-w1", "divergent")],
    expected: {
      ok: false,
      code: "writer_fork",
      details: { writerId: W, seq: 2 },
    },
  },
  {
    name: "existing entry_id with different canonical bytes → entry_id_collision (§9.3 error 3, §6.3)",
    present: [w1, w2],
    // A structurally valid extension that reuses id-w1 for different bytes.
    batch: [mk(W, 3, "id-w1", "id-w2", "reused-id")],
    expected: {
      ok: false,
      code: "entry_id_collision",
      details: { entryId: "id-w1" },
    },
  },
  {
    name: "same entry_id at the SAME coordinate with different bytes → entry_id_collision, not fork (§6.3 MUST)",
    present: [w1, w2],
    // Same id, same (writer, seq) as the stored w2, but tampered bytes.
    // Kills the mutant that drops the byte-equality clause from the
    // stored-id check: without it this row degrades to writer_fork.
    batch: [mk(W, 2, "id-w2", "id-w1", "tampered")],
    expected: {
      ok: false,
      code: "entry_id_collision",
      details: { entryId: "id-w2" },
    },
  },
  {
    name: "same entry_id, same bytes, same coordinate → duplicate, present, NOT a collision (§9.3 rule 2)",
    present: [w1, w2],
    batch: [mk(W, 2, "id-w2", "id-w1")],
    expected: { ok: true, perWriter: [], presentCount: 1 },
  },
  {
    name: "intra-batch sequence hole after a valid first new entry → writer_gap at the hole (§9.3 rule 5)",
    present: [w1, w2],
    batch: [
      mk(W, 3, "id-w3", "id-w2"),
      mk(W, 4, "id-w4", "id-w3"),
      mk(W, 6, "id-w6", "id-w5"),
    ],
    expected: {
      ok: false,
      code: "writer_gap",
      details: { writerId: W, expectedSeq: 5, tipEntryId: "id-w2" },
    },
  },
  {
    name: "intra-batch prev_entry_id not naming the preceding batch entry → invalid_batch (§9.3 rule 5)",
    present: [w1, w2],
    batch: [mk(W, 3, "id-w3", "id-w2"), mk(W, 4, "id-w4", "id-BOGUS")],
    expected: { ok: false, code: "invalid_batch" },
  },
  {
    name: "intra-batch entry_id reused at different coordinates → entry_id_collision (§6.3)",
    present: [],
    batch: [mk("wf", 1, "f-dup", null), mk("wf", 2, "f-dup", "f-dup")],
    expected: {
      ok: false,
      code: "entry_id_collision",
      details: { entryId: "f-dup" },
    },
  },
  {
    name: "intra-batch duplicate coordinate with different bytes → writer_fork",
    present: [],
    batch: [mk("wf", 1, "f1", null), mk("wf", 1, "f1b", null)],
    expected: {
      ok: false,
      code: "writer_fork",
      details: { writerId: "wf", seq: 1 },
    },
  },
  {
    name: "exact duplicate entry within one batch → deduplicated, copy counted as present",
    present: [],
    batch: [mk("wf", 1, "f1", null), mk("wf", 1, "f1", null)],
    expected: {
      ok: true,
      presentCount: 1,
      perWriter: [
        { writerId: "wf", firstNewSeq: 1, newEntries: [{ entryId: "f1" }] },
      ],
    },
  },
  {
    name: "multi-writer batch where one writer fails → whole batch fails with that writer's error, no partial plan",
    present: [w1, w2],
    batch: [
      // "aaa" sorts before W and is entirely valid; the classifier must
      // still refuse the whole batch for W's gap. ok:false carries no
      // perWriter, so no partial plan can escape (§8 invariant 10).
      mk("aaa", 1, "a1", null),
      mk(W, 4, "id-w4", "id-w3"),
    ],
    expected: {
      ok: false,
      code: "writer_gap",
      details: { writerId: W, expectedSeq: 3, tipEntryId: "id-w2" },
    },
  },
  {
    name: "multi-writer all-valid batch → one plan per writer, sorted by writerId",
    present: [w1, w2],
    batch: [
      // Deliberately fed in reverse writer order.
      mk("zzz", 1, "z1", null),
      mk(W, 3, "id-w3", "id-w2"),
      mk("aaa", 1, "a1", null),
      mk("aaa", 2, "a2", "a1"),
    ],
    expected: {
      ok: true,
      presentCount: 0,
      perWriter: [
        {
          writerId: "aaa",
          firstNewSeq: 1,
          newEntries: [{ entryId: "a1" }, { entryId: "a2" }],
        },
        { writerId: W, firstNewSeq: 3, newEntries: [{ entryId: "id-w3" }] },
        { writerId: "zzz", firstNewSeq: 1, newEntries: [{ entryId: "z1" }] },
      ],
    },
  },
];

// --- runner ----------------------------------------------------------------

describe("planMerge classifies batches per spec §9.3", () => {
  it("anti-vacuity: the table has at least 14 rows", () => {
    expect(rows.length).toBeGreaterThanOrEqual(14);
  });

  it.for(rows)("$name", (row) => {
    const { existing, presentByCoord, presentById } = storeOf(row.present);
    const outcome: MergeOutcome = planMerge(
      row.batch,
      existing,
      presentByCoord,
      presentById,
    );
    expect(outcome).toMatchObject(row.expected);
  });

  it("batch order does not matter: shuffled input classifies identically (§9.3 rule 1)", () => {
    const batch = [
      mk(W, 3, "id-w3", "id-w2"),
      mk(W, 4, "id-w4", "id-w3"),
      mk(W, 5, "id-w5", "id-w4"),
    ];
    const shuffled = [batch[2]!, batch[0]!, batch[1]!];
    const { existing, presentByCoord, presentById } = storeOf([w1, w2]);
    const a = planMerge(batch, existing, presentByCoord, presentById);
    const b = planMerge(shuffled, existing, presentByCoord, presentById);
    expect(b).toEqual(a);
    expect(a).toMatchObject({
      ok: true,
      perWriter: [{ writerId: W, firstNewSeq: 3 }],
    });
  });
});
